core/src/main/java/org/apache/calcite/rex/RexNormalize.java

```diff
@@ -0,0 +1,198 @@
+    }
+    return original;
+  }
+
+  /**
+   * Compares two operands to see which one should be normalized to be in front of the other.
+   *
+   * <p>We can always use the #hashCode to reorder the operands, do it as a fallback to keep
+   * good readability.
+   *
+   * @param operand0  First operand
+   * @param operand1  Second operand
+   *
+   * @return non-negative (>=0) if {@code operand0} should be in the front,
+   * negative if {@code operand1} should be in the front
+   */
+  private static int reorderOperands(RexNode operand0, RexNode operand1) {
+    // Reorder the operands based on the SqlKind enumeration sequence,
+    // smaller is in the behind, e.g. the literal is behind of input ref and AND, OR.
+    int x = operand0.getKind().compareTo(operand1.getKind());
+    // If the operands are same kind, use the hashcode to reorder.
+    // Note: the RexInputRef's hash code is its index.
+    return x != 0 ? x : operand1.hashCode() - operand0.hashCode();
+  }
+
+  /** Computes the hashcode of a rex call, we ignore the operands sequence when
+   * the call is symmetrical. */
+  public static int hashCode(SqlOperator operator, List<RexNode> operands) {
+    if (!allowsNormalize()) {
+      return Objects.hash(operator, operands);
+    }
+    // Quick check.
+    if (operands.size() != 2) {
+      return Objects.hash(operator, operands);
+    }
+    if (isSymmetricalCall(operator, operands.get(0), operands.get(1))) {
+      return Objects.hash(operator, unorderedHash(operands, 17));
+    }
+    // If arguments are the same, then we normalize < vs >
+    // '<' == 60, '>' == 62, so we prefer <.
+    final SqlKind kind = operator.getKind();
+    final int x = kind.reverse().compareTo(kind);
+    if (x < 0) {
+      // Note that Objects.hash([obj1, obj2]) is different from
+      // Objects.hash(obj1, obj2)
+      return Objects.hash(
+          SqlStdOperatorTable.reverse(operator),
+          ImmutableList.of(operands.get(1), operands.get(0)));
+    } else {
+      return Objects.hash(operator, operands);
+    }
+  }
+
+  /** Returns whether a call is symmetrical. **/
+  private static boolean isSymmetricalCall(
+      SqlOperator operator,
+      RexNode operand0,
+      RexNode operand1) {
+    return operator.isSymmetrical()
+        || SqlKind.SYMMETRICAL_SAME_ARG_TYPE.contains(operator.getKind())
+            && SqlTypeUtil.equalSansNullability(operand0.getType(), operand1.getType());
+  }
+
+  /**
+   * The digest of {@code RexNode} is normalized by default.
+   *
+   * @return true if the digest needs to be normalized
+   */
+  private static boolean allowsNormalize() {
+    return CalciteSystemProperty.ENABLE_REX_DIGEST_NORMALIZE.value();
+  }
+
+  /** Compute a hash that is symmetric in its arguments - that is a hash
+   *  where the order of appearance of elements does not matter.
+   *  This is useful for hashing symmetric rex calls, for example.
+   */
+  private static int unorderedHash(List<?> xs, int seed) {
```

Comment:
Could you please inline the seed into the method?
I don't think we would need to change it often, so having less unknowns in the API helps.

Reply:
By the way, I don't think `unorderedHash` helps here.
We still need to be able to implement `equals`, and I don't think there's an easy way to compute `equals` without normalization.

Then, I think we could normalize and cache the normalized value, then we can use a regular hash functions rather than `unordered` ones.

Reply:
Agree, would cache the normalized value instead.

Reply:
The thing is it is quite possible that you don't even get chance to call `equals` at all, in which case, it is wasteful to normalize it in computing hashcode.

Reply:
That makes sense!

In other words, use `unorderedHash` in `hashCode`, and compute/store the normalization only for `equals`.

Reply:
I'm fine to keep the `unorderedHash`, one drawback with it is that the logic to decide whether operands need reordering must be strictly same with the `#normalize`, which i don't like because it is error-prone.